### ISSUE TITLE
🏈NFL Helmet Assignment - Tuning DeepSort + Helmet Mapping_high_score

### ISSUE THREAD

**Issue by ryryrymyg** [ryryrymyg/kaggle_for_me#1]
# Tuning DeepSort + Helmet Mapping_high_score
DeepSortとHelmet Mappingを用いた検出方法のノート
コードをノートから貼り付けながら何をしているのか詳らかにしていく

## 第1コードブロック
```python
# Install helmet-assignment helper code
!pip install ../input/helmet-assignment-helpers/helmet-assignment-main/ > /dev/null 2>&1
from helmet_assignment.score import NFLAssignmentScorer, check_submission
from helmet_assignment.features import add_track_features 
```
ここではこのコンペ特有の関数をimportしている。
まず、pipで/helmet-assignment-main/配下のpythonファイルをインストールし、以下のファイルをimportしている。
- score.py
  - NFLAssignmentScorer
    大体のスコアを計算してくれる
  - check_submission
    提出の縛りに従っているかどうか調べてくれる
- features.py
  - add_track_features：https://www.kaggle.com/robikscube/nfl-helmet-assignment-getting-started-guide
  それぞれ別のノートブックからインポートされたもので、概要は説明に記載した通り。
    - game_play列を追加する: [game_key]_playID(6桁0左詰)
    - time列をdatetime型に変更する
    - snap_dict: event列がball_snapである列を抽出し、それらとgame_play列の最初の時間とを対応づけたdict型列
    - snap列を追加する: game_play列にsnap_dict関数を適応した列を追加する。ball_snapの最初の値
    - isSnap列を追加する。snap列とtime列が一致するかどうかの真偽値 
    - team列を追加する: player列の頭のHをHome, VをAwayに変更したもの
    - snap_offset列を追加する: ball_snapまでの時間を表す列。time列からsnap列を差し引いた値(小数点日時型(timedelta64):ms)。
    - est_frame列を追加する: ball_snapまでのフレームを見積もった列。snap_offset列にfpsをかけてroundで丸めたもの

# Baseline helmet mapping
次のノートブックからこのヘルメットマッピングのやり方は拝借している。
[https://www.kaggle.com/its7171/nfl-baseline-simple-helmet-mapping](url)

## 第2コードブロック
```python
import numpy as np
import pandas as pd
import itertools
import glob
import os
import cv2
from sklearn.metrics import accuracy_score
from tqdm.auto import tqdm
from multiprocessing import Pool
from matplotlib import pyplot as plt
from sklearn.cluster import KMeans
import random
```
ここでは今回のモデルで用いるライブラリをインポートしている。
以下で各ライブラリの簡単な説明をする
- numpy
  言わずと知れたNumpy。ndarray便利。確かpandas使うのにもいる
- pandas
  多次元配列演算のために用いるモジュール。その他データの中身を簡単に覗く際にも便利。(pandas_profiling)
- itertools
  効率的なループ生成のためのイテレータ生成ツール。とされているが、実際は実装が面倒な計算をさくっとやってくれるモジュール。累積和や組み合わせ、リスト要素の数え上げなど。numpyでもできるけどitertoolの方が都合が良い場合があるらしい。
[https://qiita.com/anmint/items/37ca0ded5e1d360b51f3](url)
[https://docs.python.org/ja/3/library/itertools.html](url)
- glob
  ファイル操作のためのモジュール
- os
  OSとのインタフェースを提供するモジュール
- cv2
  OpenCVを使うためのモジュール
  OpenCVの教科書[https://jellyware.jp/aicorex/contents/out_c04_opencv.html](url)
- accuracy_score
  2つのオブジェクトを比較して正解率(accuracy)を計算する関数
  [https://pythondatascience.plavox.info/scikit-learn/%E5%88%86%E9%A1%9E%E7%B5%90%E6%9E%9C%E3%81%AE%E3%83%A2%E3%83%87%E3%83%AB%E8%A9%95%E4%BE%A1](url)
- from tqdm.auto import tqdm
  プログレスバーをつけるためのモジュール
- from multiprocessing import Pool
  関数の並列実行を行うためのオブジェクト
  [https://docs.python.org/ja/3/library/multiprocessing.html](url)
- from matplotlib import pyplot as plt
  図表出すためのモジュール。seabornと比べると若干煩雑だけどその分カスタマイズ性がある
- from sklearn.cluster import KMeans
  クラスタ分析の中のK-Means方を用いるためのクラス
  [https://pythondatascience.plavox.info/scikit-learn/%E3%82%AF%E3%83%A9%E3%82%B9%E3%82%BF%E5%88%86%E6%9E%90-k-means](url)
- random
  乱数を得るためのモジュール

# Settings and loading data
設定とデータの読み込み
実験しやすいように、`max_iter`、`DIG_STEP`、`DIG_MAX`を一番上に抽出したことに注意する。また、パブリックテストセットで実行する場合は、デバッグモードで実行するようにコードを修正した。

## 第3コードブロック
```python
### os.listdir(): 引数のディレクトリ配下のファイルをリスト形式で出力する
n_test_videos = len(os.listdir('../input/nfl-health-and-safety-helmet-assignment/test/'))
# Run in debug mode unless during submission
if n_test_videos == 6:
    debug = True
else:
    debug = False

# Configurables以下の数値はまだ謎
n_debug_samples = 1. 
random_state = 42
CONF_THRE = 0.4
max_iter = 1000
DIG_STEP = 3
DIG_MAX = DIG_STEP*10

# Read in the data.元データのあるディレクトリの指定

BASE_DIR = '../input/nfl-health-and-safety-helmet-assignment'

labels = pd.read_csv(f'{BASE_DIR}/train_labels.csv')
if debug:
    tracking = pd.read_csv(f'{BASE_DIR}/train_player_tracking.csv')
    helmets = pd.read_csv(f'{BASE_DIR}/train_baseline_helmets.csv')
else:
    tracking = pd.read_csv(f'{BASE_DIR}/test_player_tracking.csv')
    helmets = pd.read_csv(f'{BASE_DIR}/test_baseline_helmets.csv')
    
tracking = add_track_features(tracking)
```
- n_test_videos = len(os.listdir('../input/nfl-health-and-safety-helmet-assignment/test/'))
  テストビデオの数。
- os.listdir()
  指定したディレクトリのファイル名のリスト
- labels = pd.read_csv(f'{BASE_DIR}/train_labels.csv')
  train_labels.csvをpandasのDFに変換してlabelsに代入している。f’{}’は関数を文字列に代入するための手法
- tracking
  プレイヤーがコート上のどこにいるのかのデータ
- train_baseline_helmets.csv
  images内の画像とラベルで学習したモデルが出力したデータ
- test_baseline_helmets.csv
  testデータから用意されたヘルメットのデータ？
- tracking = add_track_features(tracking)
  選手のトラッキングデータに上述のデータ列を追加する。

## 第4コードブロック
```python

def add_cols(df):
    df['game_play'] = df['video_frame'].str.split('_').str[:2].str.join('_')
    if 'video' not in df.columns:
        df['video'] = df['video_frame'].str.split('_').str[:3].str.join('_') + '.mp4'
    return df

if debug:
    helmets = add_cols(helmets)
    labels = add_cols(labels)
    # Select `n_debug_samples` worth of videos to debug with
    sample_videos = labels['video'].drop_duplicates() \
        .sample(n_debug_samples, random_state=random_state).tolist()
    sample_gameplays = ['_'.join(x.split('_')[:2]) for x in sample_videos]
    tracking = tracking[tracking['game_play'].isin(sample_gameplays)]
    helmets = helmets[helmets['video'].isin(sample_videos)]
    labels = labels[labels['video'].isin(sample_videos)]
tracking.shape, helmets.shape, labels.shape
```
- add_cols(df)
  与えられたデータフレームにgame_play列を追加する。game_play列はtrackingのgame_play列と対応する
  video列がないものに対してはvideo列を追加する。video列は与えられたvideoに対応する
- if debug以下
  testデータ群に対する処理。trainデータ群にはすでに施されている。
- sample_videos = labels['video'].drop_duplicates() .sample(n_debug_samples, random_state=random_state).tolist()
  ビデオのサンプルを一つ取る
- sample_gameplays = ['_'.join(x.split('_')[:2]) for x in sample_videos]
  サンプルの順番のgame_playsをリスト化したもの
- tracking = tracking[tracking['game_play'].isin(sample_gameplays)]
- helmets = helmets[helmets['video'].isin(sample_videos)]
- labels = labels[labels['video'].isin(sample_videos)]
  上記三つのデータについて、それぞれsampleデータのみを抽出する

## 第5コードブロック
```python

def find_nearest(array, value):
    value = int(value)
    array = np.asarray(array).astype(int)  # asarray: arrayをNumpyh配列に変換する。その際array(新)とarray(旧)は同一配列
    idx = (np.abs(array - value)).argmin() # array - valueが最小となる配列のインデックス
    return array[idx]

def norm_arr(a): # ベクトルaのノルムを計算する
    a = a-a.min()
    a = a/a.max()
    return a
    
def dist(a1, a2): # a1とa2の間の距離(ノルム)を計算する
    return np.linalg.norm(a1-a2)

def dist_for_different_len(a1, a2):
    assert len(a1) >= len(a2), f'{len(a1)}, {len(a2)}'
    len_diff = len(a1) - len(a2)
    a2 = norm_arr(a2)
    if len_diff == 0:
        a1 = norm_arr(a1)
        return dist(a1,a2), ()
    else:
        min_dist = 10000
        min_detete_idx = None
        cnt = 0
        del_list = list(itertools.combinations(range(len(a1)),len_diff))
        if len(del_list) > max_iter:
            del_list = random.sample(del_list, max_iter)
        for detete_idx in del_list:
            this_a1 = np.delete(a1, detete_idx)
            this_a1 = norm_arr(this_a1)
            this_dist = dist(this_a1, a2)
            #print(len(a1), len(a2), this_dist)
            if min_dist > this_dist:
                min_dist = this_dist
                min_detete_idx = detete_idx
                
        return min_dist, min_detete_idx
        
def rotate_arr(u, t, deg=True):
    if deg == True:
        t = np.deg2rad(t)
    R = np.array([[np.cos(t), -np.sin(t)],
                  [np.sin(t),  np.cos(t)]])
    return  np.dot(R, u)

def dist_rot(tracking_df, a2):
    tracking_df = tracking_df.sort_values('x')
    x = tracking_df['x']
    y = tracking_df['y']
    min_dist = 10000
    min_idx = None
    min_x = None
    for dig in range(-DIG_MAX,DIG_MAX+1,DIG_STEP):
        arr = rotate_arr(np.array((x,y)), dig)
        this_dist, this_idx = dist_for_different_len(np.sort(arr[0]), a2)
        if min_dist > this_dist:
            min_dist = this_dist
            min_idx = this_idx
            min_x = arr[0]
    tracking_df['x_rot'] = min_x
    player_arr = tracking_df.sort_values('x_rot')['player'].values
    players = np.delete(player_arr,min_idx)
    return min_dist, players


def mapping_df(args):
    video_frame, df = args
    gameKey,playID,view,frame = video_frame.split('_')
    gameKey = int(gameKey)
    playID = int(playID)
    frame = int(frame)
    this_tracking = tracking[(tracking['gameKey']==gameKey) & (tracking['playID']==playID)]
    est_frame = find_nearest(this_tracking.est_frame.values, frame)
    this_tracking = this_tracking[this_tracking['est_frame']==est_frame]
    len_this_tracking = len(this_tracking)
    df['center_h_p'] = (df['left']+df['width']/2).astype(int)
    df['center_h_m'] = (df['left']+df['width']/2).astype(int)*-1
    df = df[df['conf']>CONF_THRE].copy()
    if len(df) > len_this_tracking:
        df = df.tail(len_this_tracking)
    df_p = df.sort_values('center_h_p').copy()
    df_m = df.sort_values('center_h_m').copy()
    
    if view == 'Endzone':
        this_tracking['x'], this_tracking['y'] = this_tracking['y'].copy(), this_tracking['x'].copy()
    a2_p = df_p['center_h_p'].values
    a2_m = df_m['center_h_m'].values

    min_dist_p, min_detete_idx_p = dist_rot(this_tracking ,a2_p)
    min_dist_m, min_detete_idx_m = dist_rot(this_tracking ,a2_m)
    if min_dist_p < min_dist_m:
        min_dist = min_dist_p
        min_detete_idx = min_detete_idx_p
        tgt_df = df_p
    else:
        min_dist = min_dist_m
        min_detete_idx = min_detete_idx_m
        tgt_df = df_m
    #print(video_frame, len(this_tracking), len(df), len(df[df['conf']>CONF_THRE]), this_tracking['x'].mean(), min_dist_p, min_dist_m, min_dist)
    tgt_df['label'] = min_detete_idx
    return tgt_df[['video_frame','left','width','top','height','label']]

p = Pool(processes=4)
submission_df_list = []
df_list = list(helmets.groupby('video_frame'))
with tqdm(total=len(df_list)) as pbar:
    for this_df in p.imap(mapping_df, df_list):
        submission_df_list.append(this_df)
        pbar.update(1)
p.close()

submission_df = pd.concat(submission_df_list)
submission_df.to_csv('submission-baseline.csv', index=False)
```

【わからんこと】
- test_baseline_helmets.csv
  これ何に使うんや
- *_player_tracking.csv
  eventのball_snapってなんや
- timedelta64型
  日時の少数点型？1_000ってなんや
【ミーティングで聞くこと】
- 関数、メソッド、ノートブックごとのISSUEっている？
  いらなければ個人的なISSUEにまとめる
